### PR TITLE
Handle 404, 500 & 502 status codes

### LINKED ISSUE #1214: Handle 404, 500 & 502 appropriately - server-side
**Is your feature request related to a problem? Please describe.**
Technical requirements for Article pages:

* Ares throws 404 -> Simorgh throws 404
* Ares throws anything except 200 and 404 -> Simorgh throws 502
* Simorgh has internal error -> Simorgh throws 500
* Simorgh unable to parse corrupt Ares data -> Simorgh throws 502

This issue is solely to do with the response code Simorgh throws when faced with unusual situations. It is NOT about what the error pages looks like.

**Describe the solution you'd like**
Currently, we serve a 404 for any error that is thrown in our application. This should be a 500. 
https://github.com/bbc/simorgh/blob/939c99100eef177dde725bf293bd37e20f939551/src/server/index.jsx#L95

Throw a 500 error here in the Article container: https://github.com/bbc/simorgh/blob/939c99100eef177dde725bf293bd37e20f939551/src/app/containers/Article/index.jsx#L61

Throw a 502 error if Ares returns a 500. 

**Describe alternatives you've considered**

**Testing notes**
Cypress tests to be added - pairing with a developer. 

Dev insight: We can use `supertest` (as we currently do) to check response codes. This lives in `server.index.test.jsx`. We should ensure we have enough unit tests to cover all of the scenarios above.

**Additional context**
Error page design implementation is covered by another issue: #1194

- [x] Initially labelled with ["Refinement needed"](https://github.com/bbc/simorgh/labels/Refinement%20Needed)

### PR DESCRIPTION
**Pull request by bcmn**
Resolves #1214

**Overall change:** Ensures Simorgh responds with suitable error codes, as defined by the following rules:
- If the data source 404s, Simorgh 404s
- If the data source responds with any non-200, non-404 status code, Simorgh 502s
- If the data source 200s, but responds with invalid JSON, Simorgh 502s
- If Simorgh throws an unhandled error, Simorgh 500s

**Code changes:**

- Changes `getInitialData` to return a status code, & ensure it only ever returns one of `200`, `404` or `502`. If the data source returns anything else, it will be logged & `getInitialData` will instead return `502`. The solution I ended up with here feels less intimidating than an entirely new layer handling the mapping, but perhaps this could be re-evaluated.
- Refactors `ArticleContainer`, to provide a consistent page skeleton & avoid any assumptions that the article data will always be present. This helps set us up for error & loading pages, which will require a large amount of shared page furniture that was previously entangled in logic.
- The business logic that previously lived in `ArticleContainer` now lives in the `ArticleMain` container. `ArticleContainer` now only branches on the `status` it receives from `getInitialData`, choosing different `main` containers to switch in.
- This PR introduces `ErrorMain` & passes it `status`, providing a basis for the error page work coming up. Currently, it looks like this:
screenshot 2019-02-06 at 16 06 17
- Finally, `server/index.jsx` now responds with the status code determined by `getInitialData`. As described above, it feels a little strange for our data fetcher to also determine _our_ status code, but without it there would need to be something more complex, that still allows for a solution that works across both server & client.

Small refactors: 
- moves `articleDataRegexPath` so all our path regexes live in `routes.js`
- renames the Cypress test helper `testNonHtmlResponseCode` to `testResponseCode`
- changes MetadataContainer to utilise ServiceContextProvider when creating canonical links, removing the need for a `service` prop in both ArticleMain & MetadataContainer

---

- [x] I have assigned myself to this PR and the corresponding issues
- [ ] Tests added for new features
- [ ] Test engineer approval
- [ ] I have followed [the merging checklist](https://github.com/bbc/simorgh-infrastructure/blob/latest/documentation/MERGE_PROCESS.md) and this is ready to merge.

### CODE REVIEW

File: src/app/containers/ArticleMain/index.jsx [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+const ArticleMain = ({ service, articleData }) => {
+  const { content, metadata, promo } = articleData;
+  const { headlineBlocks, mainBlocks } = splitBlocksByHeadline(content);
+
+  /*
+   * headlineBlocks length check is temporary
+   * Simorgh will respond with 400 to lack of headline block in issue
+   * https://github.com/bbc/simorgh/issues/836
+   */
+  if (headlineBlocks.length > 0) {
```
Comment (bcmn) [inline src/app/containers/ArticleMain/index.jsx:43]: I've kept this logic as altering it seems out of scope for this PR. However, I'm not sure we intend to enforce it any more, & even if we do, returning 4XX seems incorrect -- this isn't a user/request error.

If we wanted to enforce it, following this PR we should be able to simply throw an error & return a 500.

**Comment by bcmn** [comment]
A handful of Code Climate issues, but they're all to do with repeating setups for tests. 🤷‍♂️

File: src/app/routes/getInitialData/index.js [hunk 2]
```diff
@@ -1,27 +1,40 @@
 import 'isomorphic-fetch';
 
+const upstreamStatusCodesToPropagate = [200, 404, 502];
```
Comment (dr3) [inline src/app/routes/getInitialData/index.js]:
```suggestion
const upstreamStatusCodesToPropagate = [200, 404];
```
Keeps the added benefit of the `Unexpected upstream response` console warning

**Review by dr3 — CHANGES_REQUESTED** [review]
Really good work, initial look over but looks brilliant :D

File: src/app/routes/getInitialData/index.js (hunk 2 above)
Comment (bcmn) [reply src/app/routes/getInitialData/index.js]: Good shout. Thanks!

File: src/server/index.jsx [hunk 3]
```diff
@@ -73,32 +99,17 @@ server
       }
     });
   })
-  .get('/*', async ({ url }, res) => {
+  .get(articleRegexPath, async ({ url }, res) => {
```
Comment (dr3) [inline src/server/index.jsx:122]: Im not convinced by this, unless its only short term untill we properly handle unmatched route error handling

**Review by dr3 — DISMISSED** [review]
Re: the express routing, any chance you could make an issue/attach to the move RUA issue to ensure we handle the routing error please?

File: cypress/integration/ampSpec.js [hunk 4]
```diff
@@ -18,18 +15,10 @@ describe('AMP Tests on a .amp page', () => {
...
-    retrieve404BodyResponse(
-      '/news/lol/c85pqyj5m2ko.amp',
-      'No route was found for /news/lol/c85pqyj5m2ko.amp.',
-    );
-    retrieve404BodyResponse(
-      '/cake/articles/c85pqyj5m2ko.amp',
-      'No route was found for /cake/articles/c85pqyj5m2ko.amp.',
-    );
+    testNonHTMLResponseCode('/news/articles/c85pqyj5m2ko.cake', 404);
+    testNonHTMLResponseCode('/news/articles/c85pqyj5m2ko.cake', 404);
```
Comment (pjlee11) [inline cypress/integration/ampSpec.js]:
```suggestion
```

duplicate of above line

File: src/app/containers/Article/__snapshots__/index.test.jsx.snap [hunk 5]
```diff
@@ -1,66 +1,21 @@
...
 
-exports[`ArticleContainer Component no data should render correctly 1`] = `null`;
-
-exports[`ArticleContainer Component should render Persian article correctly 1`] = `
+exports[`ArticleContainer Component 200 status code should render correctly 1`] = `
 <React.Fragment>
   <GlobalStyleComponent />
   <ServiceContextProvider
-    service="persian"
+    service="news"
```
Comment (pjlee11) [inline src/app/containers/Article/__snapshots__/index.test.jsx.snap:7]: We have seemingly lost the persian article render as part of these changes, before it add a news and a persian render now it only has news. Could you re-add please

File: src/app/containers/Article/index.test.jsx [hunk 6]
```diff
@@ -1,39 +1,40 @@
 import React from 'react';
 import { shouldShallowMatchSnapshot } from '../../helpers/tests/testHelpers';
 import ArticleContainer from './index';
-import { articleDataNews, articleDataPersian } from './fixtureData';
+import { articleDataNews } from './fixtureData';
```
Comment (pjlee11) [inline src/app/containers/Article/index.test.jsx:4]: We should keep `articleDataPersian` as it's proof that the ServiceContext is working correctly in regards to the snapshots

File: src/app/containers/Article/__snapshots__/index.test.jsx.snap (hunk 5 above)
Comment (pjlee11) [reply src/app/containers/Article/__snapshots__/index.test.jsx.snap:7]: I've just notice that `ArticleMain` uses the persian data. However, I still think it' beneficial to have Persian data at this level as it still includes the metaData render, whereas the `ArticleMain` snapshots do not

File: src/app/containers/ArticleMain/index.jsx [hunk 7]
```diff
@@ -0,0 +1,82 @@
+import React, { Fragment } from 'react';
+import { string, shape } from 'prop-types';
+import { articleDataPropTypes } from '../../models/propTypes/article';
+import MetadataContainer from '../Metadata';
```
Comment (pjlee11) [inline src/app/containers/ArticleMain/index.jsx:4]: This is the right place but still feels weird to me. We have the metadata container in `ArticleMain` but the snapshots for `ArticleMain` do not output the metadata values. Think this is something to discuss and understand better...

File: src/app/routes/getInitialData/index.js [hunk 8]
```diff
@@ -1,27 +1,40 @@
 import 'isomorphic-fetch';
 
+const upstreamStatusCodesToPropagate = [200, 404];
+
 const getInitialData = async ({ match }) => {
-  try {
-    const { id, service, amp } = match.params;
+  const { id, service, amp } = match.params;
```
Comment (pjlee11) [inline src/app/routes/getInitialData/index.js:9]: In the case that `params` is undefined (EG: match is empty) we would fail with "Cannot read X of undefined" for `id`, `service`, `amp`. Now that this is outside of the `try` block the application could error to the user for the given case.

File: src/app/routes/getInitialData/index.js [hunk 9]
```diff
@@ -1,27 +1,40 @@
 import 'isomorphic-fetch';
 
+const upstreamStatusCodesToPropagate = [200, 404];
```
Comment (pjlee11) [inline src/app/routes/getInitialData/index.js:6]: `statusCodesToPropagateUpstream` - maybe?

Sounds a little less yoda like ...

File: src/app/routes/getInitialData/index.test.js [hunk 10]
```diff
@@ -77,8 +88,72 @@ describe('getInitialData', () => {
...
+        isAmp: false,
+        service: 'news',
+        status: 404,
+      });
+    });
+  });
+
+  describe('Ares returns a non-200, non-404 status code', () => {
+    it('should log, and return the status code as 502', async () => {
+      global.console.warn = jest.fn();
```
Comment (pjlee11) [inline src/app/routes/getInitialData/index.test.js]: A nice to have, but we could have a test for `global.console.warn.totoHaveBeenCalledWith(...)` - would make turning this into logging much easier as the test would already exist

File: src/server/index.jsx (hunk 3 above)
Comment (pjlee11) [reply src/server/index.jsx:122]: @dr3 could you open an issue to re-evaluate this in the future, can be assessed after we launch v1.0

**Review by pjlee11 — CHANGES_REQUESTED** [review]
Brilliant work so far, I've added a few comments for changes and some points for discussion

File: src/app/containers/Article/index.test.jsx (hunk 6 above)
Comment (pjlee11) [reply src/app/containers/Article/index.test.jsx:4]: Also see the MetaData point in https://github.com/bbc/simorgh/pull/1244#discussion_r254813161

File: src/app/routes/getInitialData/index.js (hunk 9 above)
Comment (bcmn) [reply src/app/routes/getInitialData/index.js:6]: "Upstream" would generally be viewed as the data source, I think? I'm saying "of the status codes the data source (upstream) returns, only propagate these". 🤷‍♂️

File: src/app/routes/getInitialData/index.js (hunk 8 above)
Comment (bcmn) [reply src/app/routes/getInitialData/index.js:9]: If included in the current try-catch, it would have to be handled in the same way as all the data fetching & basic parsing to an object below. The latter should result in a 502. 

If `match` is undefined, it suggests a fundamental problem in our configuration in `routes.js`, so a 500 would be more appropriate. Currently this is achieved [here](https://github.com/BBC-News/simorgh/blob/bb1aa9f979391daf6c533b065380b388178d60e1/src/server/index.jsx#L103-L104). 

I guess this leaves the client-side somewhat unprepared to handle it (though it's worth noting this issue was in theory scoped to the server-side), but the scenario seems very odd in the first place. I could add a `if (match == undefined) { return { isAmp: undefined, data: undefined, service: undefined, status: 500 }}` but I'm not convinced this leaves the client in any better position to know what to display -- What branding do we give the error page? What language do we present it in? If we default to specifics for these kind of questions (i.e. "just assume it's english news"), is here really the best place for that?

Comment (bcmn) [reply src/app/routes/getInitialData/index.js:9]: Perhaps:

`if (match == undefined) { return null }`?

I suspect this will trigger Uni's `error` scenario, allowing us to handle it in future wherever that gets handled. Edit: I'd need to test this.

File: src/app/containers/Article/index.test.jsx (hunk 6 above)
Comment (bcmn) [reply src/app/containers/Article/index.test.jsx:4]: In theory, the body of the article would be tested in both English & Persian in the ArticleMain snapshots -- this test now just checks if ArticleMain is called as expected. I don't think the service provider was tested before, because the ArticleContainer tests aren't deep renders -- I'm not sure they ever meet a consumer.

I hadn't noticed the metadata issue -- that somewhat undermines the idea above. I'll have to look into it.

File: src/app/routes/getInitialData/index.js (hunk 8 above)
Comment (bcmn) [reply src/app/routes/getInitialData/index.js:9]: > I suspect this will trigger Uni's `error` scenario, allowing us to handle it in future wherever that gets handled. Edit: I'd need to test this.

This is not true. It comes through as `data` & throws at the branch between ArticleMain & ErrorMain.

File: src/app/containers/Article/index.test.jsx (hunk 6 above)
Comment (bcmn) [reply src/app/containers/Article/index.test.jsx:4]: Hm, I think Metadata only showed up in ArticleContainer because it's a shallow snapshot. 😕 Maybe I do the same for ArticleMain, so long as it doesn't feel like we're losing anything from the deep render.

**Comment by bcmn** [comment]
> @dr3: Re: the express routing, any chance you could make an issue/attach to the move RUA issue to ensure we handle the routing error please?

Just to document: this is now #1253

File: src/app/containers/Article/index.test.jsx (hunk 6 above)
Comment (bcmn) [reply src/app/containers/Article/index.test.jsx:4]: Switched back to shallow rendering, which ensures Metadata container is called. Don't know how to address the ServiceContextProvider point -- as I said above, I'm not sure this was being tested before.

Comment (sareh) [reply src/app/containers/Article/index.test.jsx:4]: In this container we're implementing the change between different service data, so it would be useful to still have a test for persian as well as News. e.g. the single `200 status code should render correctly` could become `200 status code should render News article correctly` and can add a `200 status code should render Persian article correctly`. 

You should be able to have this work by changing the object `goodData` below to be called `newsData200response` and also adding a `persianData200response` which can be passed into the ArticleContainer in a unit test.

File: cypress/integration/ampSpec.js [hunk 11]
```diff
@@ -18,18 +15,9 @@ describe('AMP Tests on a .amp page', () => {
...
-    );
-    retrieve404BodyResponse(
-      '/news/lol/c85pqyj5m2ko.amp',
-      'No route was found for /news/lol/c85pqyj5m2ko.amp.',
-    );
-    retrieve404BodyResponse(
-      '/cake/articles/c85pqyj5m2ko.amp',
-      'No route was found for /cake/articles/c85pqyj5m2ko.amp.',
-    );
+    testNonHTMLResponseCode('/news/articles/c85pqyj5m2ko.cake', 404);
```
Comment (ChrisBAshton) [inline cypress/integration/ampSpec.js]: This method name reads weirdly to me.

```suggestion
    testResponseCode('/news/articles/c85pqyj5m2ko.cake', 404);
```

File: src/app/routes/getInitialData/index.test.js [hunk 12]
```diff
@@ -77,8 +88,72 @@ describe('getInitialData', () => {
...
+
+      const response = await callGetInitialData(
+        defaultContext,
+        mockFetchTeapot,
```
Comment (ChrisBAshton) [inline src/app/routes/getInitialData/index.test.js]: What's this teapot thing?

**Review by ChrisBAshton — COMMENTED** [review]
Quick review, as others have said, we should retain some tests for Persian 👍

File: cypress/integration/ampSpec.js (hunk 11 above)
Comment (bcmn) [reply cypress/integration/ampSpec.js]: I also felt this, but it wasn't a new function. What do you think, @jamesbrumpton?

File: src/app/routes/getInitialData/index.test.js (hunk 12 above)
Comment (bcmn) [reply src/app/routes/getInitialData/index.test.js]: [I needed an unexpected status code for the upstream to return](https://httpstatuses.com/418). I'll try & clarify the identifiers.

File: cypress/integration/ampSpec.js (hunk 11 above)
Comment (jamesbrumpton) [reply cypress/integration/ampSpec.js]: Agreed. It's not very explicit. Will look over this with @bcmn .

**Comment by bcmn** [comment]
> @ChrisBAshton: as others have said, we should retain some tests for Persian 👍

The existing tests for persian have simply moved to `ArticleMain.test.jsx`. 

In terms of testing the ServiceContextProvider, I was hoping for opinions on how that should be done, & clarity whether we really think that's in scope for this PR, as it wasn't done previously. Shallow rendering will never test that, even if I re-add a persian test to `ArticleContext.test.jsx`. & generally I think we avoid deep rendering in places like ArticleContainer, as it'll make for very fragile tests.

File: cypress/integration/ampSpec.js (hunk 11 above)
Comment (bcmn) [reply cypress/integration/ampSpec.js]: Updated.

File: src/app/routes/getInitialData/index.test.js (hunk 12 above)
Comment (bcmn) [reply src/app/routes/getInitialData/index.test.js]: Updated.

File: src/app/containers/Article/index.test.jsx (hunk 6 above)
Comment (sareh) [reply src/app/containers/Article/index.test.jsx:4]: Following @bcmn's comment at the bottom of the PR and since the tests have been moved to the ArticleMain, I'm happy with addressing testing context throughout the application more holistically in another PR, rather than in this error status codes PR.

**Review by sareh — APPROVED** [review]
Great work, Ben! 👍

**Review by ChrisBAshton — APPROVED** [review]

File: src/app/containers/Article/index.test.jsx (hunk 6 above)
Comment (pjlee11) [reply src/app/containers/Article/index.test.jsx:4]: Agreed with this, can we have an issue opened please

File: src/app/routes/getInitialData/index.js (hunk 9 above)
Comment (pjlee11) [reply src/app/routes/getInitialData/index.js:6]: Hmmm, still think this could be better named but I'm not sure how 🤔

File: src/app/routes/getInitialData/index.js (hunk 8 above)
Comment (pjlee11) [reply src/app/routes/getInitialData/index.js:9]: That's aggravating. Could you open an issue to address this in the future because it's not completely related to this PR but is a fall out of this PR

**Review by pjlee11 — APPROVED** [review]
Great work!

**Comment by jamesbrumpton** [comment]
LGTM
